File: docs/examples/rbac/python/test_app.py
from pathlib import Path


def test_policy(
    test_oso, ariana, bhavik, affine_types, allocator, bubble_sort, benchmarks
):
    test_oso.load_files([Path(__file__).parent / "main.polar"])

    assert test_oso.is_allowed(ariana, "read", affine_types)
    assert test_oso.is_allowed(ariana, "push", affine_types)
    assert test_oso.is_allowed(ariana, "read", allocator)
    assert test_oso.is_allowed(ariana, "push", allocator)
    assert not test_oso.is_allowed(ariana, "read", bubble_sort)
    assert not test_oso.is_allowed(ariana, "push", bubble_sort)
    assert not test_oso.is_allowed(ariana, "read", benchmarks)
    assert not test_oso.is_allowed(ariana, "push", benchmarks)

    assert not test_oso.is_allowed(bhavik, "read", affine_types)
    assert not test_oso.is_allowed(bhavik, "push", affine_types)
    assert not test_oso.is_allowed(bhavik, "read", allocator)
    assert not test_oso.is_allowed(bhavik, "push", allocator)
    assert test_oso.is_allowed(bhavik, "read", bubble_sort)
    assert not test_oso.is_allowed(bhavik, "push", bubble_sort)
    assert test_oso.is_allowed(bhavik, "read", benchmarks)
    assert test_oso.is_allowed(bhavik, "push", benchmarks)

File: docs/examples/rbac/java/TestApp.java
import com.osohq.oso.Oso;
import com.osohq.oso.Exceptions;

public class TestApp {
  public static void testPolicy() throws Exception {
    Oso oso = App.setupOso();

    App.Organization alphaAssociation = new App.Organization("Alpha Association");
    App.Organization betaBusiness = new App.Organization("Beta Business");

    App.Repository affineTypes = new App.Repository("Affine Types", alphaAssociation);
    App.Repository allocator = new App.Repository("Allocator", alphaAssociation);
    App.Repository bubbleSort = new App.Repository("Bubble Sort", betaBusiness);
    App.Repository benchmarks = new App.Repository("Benchmarks", betaBusiness);

    App.User ariana = new App.User("Ariana");
    App.User bhavik = new App.User("Bhavik");

    ariana.assignRoleForResource("owner", alphaAssociation);
    bhavik.assignRoleForResource("contributor", bubbleSort);
    bhavik.assignRoleForResource("maintainer", benchmarks);

    oso.authorize(ariana, "read", affineTypes);
    oso.authorize(ariana, "push", affineTypes);
    oso.authorize(ariana, "read", allocator);
    oso.authorize(ariana, "push", allocator);
    try { oso.authorize(ariana, "read", bubbleSort); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(ariana, "push", bubbleSort); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(ariana, "read", benchmarks); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(ariana, "push", benchmarks); } catch(Exceptions.NotFoundException e) {}

    try { oso.authorize(bhavik, "read", affineTypes); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(bhavik, "push", affineTypes); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(bhavik, "read", allocator); } catch(Exceptions.NotFoundException e) {}
    try { oso.authorize(bhavik, "push", allocator); } catch(Exceptions.NotFoundException e) {}
    oso.authorize(bhavik, "read", bubbleSort);
    try { oso.authorize(bhavik, "push", bubbleSort); } catch(Exceptions.ForbiddenException e) {}
    oso.authorize(bhavik, "read", benchmarks);
    oso.authorize(bhavik, "push", benchmarks);
  }

  public static void main(String[] args) throws Exception {
    testPolicy();
    System.out.println("Java RBAC tests pass!");
  }
}
